segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
-                            }
-                            return null;
-                        }, executor),
-                executor);
-        return loopFuture;
+    CompletableFuture<Void>  addChunkToGarbage(long transactionId, String chunkToDelete, long startTime, int attempts) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunkToDelete, startTime, attempts, TaskInfo.DELETE_CHUNK, transactionId))
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_CHUNK_QUEUED.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
-    private boolean canRun() {
-        return isRunning() && getStopException() == null && !closed.get();
+    CompletableFuture<Void>  addSegmentToGarbage(long transactionId, String segmentToDelete) {
+        if (null != taskQueue) {
+            val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
+            return taskQueue.addTask(taskQueueName, new TaskInfo(segmentToDelete, startTime, 0, TaskInfo.DELETE_SEGMENT, transactionId))
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_SEGMENT_QUEUED.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
-    /**
-     * Sets whether background cleanup is suspended or not.
-     *
-     * @param value Boolean indicating whether to suspend background processing or not.
-     */
-    void setSuspended(boolean value) {
-        suspended.set(value);
+    CompletableFuture<Void>  addSegmentToGarbage(TaskInfo taskInfo) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(taskQueueName, taskInfo)
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_SEGMENT_QUEUED.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
-    /**
-     * Adds given chunks to list of garbage chunks.
-     *
-     * @param chunksToDelete List of chunks to delete.
-     */
-    void addToGarbage(Collection<String> chunksToDelete) {
-        val currentTime = currentTimeSupplier.get();
-
-        chunksToDelete.forEach(chunkToDelete -> addToGarbage(chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0));
+    CompletableFuture<Void>  trackNewChunk(long transactionId, String chunktoTrack) {
+        if (null != taskQueue) {
+            val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunktoTrack, startTime, 0, TaskInfo.DELETE_CHUNK, transactionId))
```

Comment:
The method is about a `trackNewChunk` but the `TaskInfo` is `TaskInfo.DELETE_CHUNK`. Is that right?

Reply:
This is correct. 

I'll add a comment